### ISSUE TITLE
BOOST_ROOT

### ISSUE THREAD

**Issue by rhd** [mesonbuild/meson#669]
Hey Jussi,

Setting BOOST_ROOT in the env doesn't seem to work.

meson finds the specified modules.

`Dependency Boost (system) found: YES (1.60, /usr/local/boost_1_60_0/)`

But fails to include them on the link line.

`FAILED: app 
g++-4.8.1    -o app <list of objects> '-Wl,--no-undefined' 'foo.a' '-pthread' '-L/usr/local/boost_1_60_0/lib' '-lrt' '-L/usr/local/boost_1_60_0/lib'`

I had to fall back to using `find_library`.

meson version: `0.34.0.dev1`

**Comment by nirbheek** [comment]
Is this still broken?

**Comment by rhd** [comment]
It seems to be working now with `meson 0.39.1`. :+1:

**Comment by rhd** [comment]
Although this is a slightly different test setup. Original bug report was gcc 4.8.1 on redhat 6. My current test was gcc 4.8.4 on redhat 7.

I'm going to close this.

**Comment by rhd** [comment]
@nirbheek I'm back on the box with the original issue - redhat 6.8 with gcc 4.8.1 and this issue is still there. I'm reopening. meson is not appending the appropriate boost libs with `-l`.

**Comment by rhd** [comment]
$ ./meson/meson.py --version
0.41.1

**Comment by rhd** [comment]
```
Dependency Boost (system, filesystem, log, iostreams, chrono, program_options, thread, date_time, regex, context, coroutine) found: YES 1.60, /usr/local/boost_1_60_0/
```

results in this `gcc` command

```
ccache g++-4.8.1  -o apps/scatterer_gatherer/scatterer 'apps/scatterer_gatherer/scatterer@exe/Scatterer_main.cpp.o' -Wl,--no-undefined -Wl,--as-needed -Wl,--start-group libs/MSAUtil/libMSAUtil.a -Wl,--end-group -pthread -L/usr/local/boost_1_60_0/lib -L/usr/local/boost_1_60_0/lib  
```

**Comment by rhd** [comment]
`meson-log.txt` is filled with errors like this:

```
Running compile:
Working directory:  /tmp/tmp8yks6kom
Command line:  ccache g++-4.8.1 /tmp/tmp8yks6kom/testfile.cpp -O0 -fpermissive -Wl,--start-group -lboost_regex -Wl,--end-group -pipe -D_FILE_OFFSET_BITS=64 -o /tmp/tmp8yks6kom/output.exe 

Code:
 int main(int argc, char **argv) {
    return 0;
}
        
Compiler stdout:
 
Compiler stderr:
 /usr/bin/ld: cannot find -lboost_regex
collect2: error: ld returned 1 exit status
```

Doesn't seem to be following the `BOOST_ROOT` at all.

**Comment by rhd** [comment]
I poked around and this patch seems to fix the issue (patched for debug purposes only)

```diff
diff --git a/meson/mesonbuild/dependencies/misc.py b/meson/mesonbuild/dependencies/misc.py
index c24acf0..cb11657 100644
--- a/meson/mesonbuild/dependencies/misc.py
+++ b/meson/mesonbuild/dependencies/misc.py
@@ -58,6 +58,7 @@ class BoostDependency(ExternalDependency):
                     self.incdir = '/usr/include'
         else:
             self.incdir = os.path.join(self.boost_root, 'include')
+            self.libdir = os.path.join(self.boost_root, 'lib')
         self.boost_inc_subdir = os.path.join(self.incdir, 'boost')
         mlog.debug('Boost library root dir is', self.boost_root)
         self.src_modules = {}
@@ -239,7 +240,7 @@ class BoostDependency(ExternalDependency):
             module = BoostDependency.name2lib.get(module, module)
             libname = 'boost_' + module
             # The compiler's library detector is the most reliable so use that first.
-            default_detect = self.compiler.find_library(libname, self.env, [])
+            default_detect = self.compiler.find_library(libname, self.env, self.libdir)
             if default_detect is not None:
                 if module == 'unit_testing_framework':
                     emon_args = self.compiler.find_library('boost_test_exec_monitor')
```

**Comment by nirbheek** [comment]
This is fixed with #2282
